src/plugins/mer/merbuildconfiguration.cpp

```diff
@@ -0,0 +1,143 @@
+
+#include <utils/algorithm.h>
+#include <utils/qtcassert.h>
+
+#include <QTextStream>
+#include <QRegularExpression>
+#include <QDebug>
+
+using namespace Utils;
+using namespace ProjectExplorer;
+using namespace QmakeProjectManager;
+
+namespace {
+
+const char MER_VARIABLES_CACHE_FILENAME[] = "/.qmake_variables.cache";
+
+EnvironmentItem parseVariableString(const QString &line)
+{
+    const QChar sep = '=';
+    QStringList chunks = line.split(sep);
+    if (chunks.size() < 2)
+        return EnvironmentItem(QString(), QString());
+
+    const QString key = chunks.takeFirst();
+    const QString value = chunks.join(sep);
+
+    return EnvironmentItem(key, value);
+}
+
+} // namespace anonymous
+
+namespace Mer {
+namespace Internal {
+
+MerBuildConfiguration::MerBuildConfiguration(Target *target, Core::Id id)
+    : QmakeBuildConfiguration(target, id)
+    , m_cacheFileName(buildDirectory().toString() + MER_VARIABLES_CACHE_FILENAME)
+{
+    initBuildEngineVariables(m_cacheFileName);
+
+    QmakeProject *qmakeProject = static_cast<QmakeProject *>(project());
+    connect(qmakeProject, &QmakeProject::buildDirectoryInitialized,
+            this, &MerBuildConfiguration::initCacheWatcher);
+
+    connect(this, &BuildConfiguration::buildDirectoryChanged, [this]() {
+        m_cacheFileName = buildDirectory().toString() + MER_VARIABLES_CACHE_FILENAME;
+        if (m_cacheWatcher.files().size())
+            m_cacheWatcher.removePaths(m_cacheWatcher.files());
+
+        initCacheWatcher();
+    });
+
+    connect(&m_cacheWatcher, &QFileSystemWatcher::fileChanged,
+            this, &MerBuildConfiguration::initBuildEngineVariables);
+}
+
+void MerBuildConfiguration::initCacheWatcher()
+{
+    if (QFile(m_cacheFileName).exists()
+            && !m_cacheWatcher.files().contains(m_cacheFileName)) {
+        m_cacheWatcher.addPath(m_cacheFileName);
+        initBuildEngineVariables(m_cacheFileName);
+    }
+}
+
+void MerBuildConfiguration::useBuildEngineVariables(bool value)
+{
+    m_useBuildEngineVariables = value;
+    setUserEnvironmentChanges(m_useBuildEngineVariables
+                              ? cleanUserEnvironmentChanges() + m_buildEngineVariables
+                              : cleanUserEnvironmentChanges());
+}
+
+void MerBuildConfiguration::initBuildEngineVariables(const QString &fileName)
+{
+    m_buildEngineVariables.clear();
+
+    QFile file(fileName);
+    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        QTextStream in(&file);
+        while (!in.atEnd()) {
+            const QString line = in.readLine();
+            const EnvironmentItem item = parseVariableString(line);
+            if ((item.name.isNull() && item.value.isNull())
+                    || m_buildEngineVariables.contains(item))
+                continue;
+
+            m_buildEngineVariables << item;
+        }
+    }
+
+    if (m_useBuildEngineVariables)
+        setUserEnvironmentChanges(cleanUserEnvironmentChanges() + m_buildEngineVariables);
```

Comment:
This mixes environment variables with qmake variables, which may lead to surprising results, very hard to debug for someone unaware of this hack. I would prefer finding different approach. Ideally some that would directly influence the .pro parser, not propagating to the actual qmake call.

Reply:
I think your approach is possible. One thing: pro parser is inside QtCreator source code. I will have to make changes there. What about upstream integration in this case?

Reply:
Yup, if we are sure that there is no existing API that could be used for that, we can modify upstream code, in which case we should ideally get the changes integrated by upstream before we merge them here. Let's see if/what changes are needed.